### PR TITLE
TF safetensors reduced mem usage

### LINKED ISSUE #24393: Increased peak memory usage when upgrading to `transformers` v4.30 and inclusion of `safetensors`
### System Info

- `transformers` version: 4.30.2
- Platform: Linux-5.19.0-1022-gcp-x86_64-with-glibc2.35
- Python version: 3.10.6
- Huggingface_hub version: 0.15.1
- Safetensors version: 0.3.1
- PyTorch version (GPU?): not installed (NA)
- Tensorflow version (GPU?): 2.12.0 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@gante @Rocketknight1 

From June 8, after upgrading to `transformers` version 4.30 which automatically installs `safetensors`, peak memory usage in BertLarge and T5Large (and possibly other models that we have not measured) increased to what appears to be a fixed value for smaller batch sizes.

image

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Run the commands below on a CUDA enabled Linux machine:

```
# Clone benchmarking repo
git clone https://github.com/iree-org/iree-samples.git
cd iree-samples/iree-tf/benchmark

# Setup output file.
OUTPUT_PATH=/tmp/tf.json
echo "{\"trigger\": { \"timestamp\": \"$(date +'%s')\" }, \"benchmarks\": []}" > "${OUTPUT_PATH}"

# Setup virtual environment.
TENSORFLOW_VERSION=2.12.0 VENV_DIR=tf.venv ./setup_venv.sh
source tf.venv/bin/activate

# Run benchmark.
BENCHMARK_ID=47cb0d3a-5eb7-41c7-9d7c-97aae7023ecf-MODEL_BERT_LARGE-fp32-TF-384xi32-batch1
python benchmark_model.py --benchmark_id="${BENCHMARK_ID}" --device=gpu --output_path="${OUTPUT_PATH}" --iterations=5
```

Benchmark output will show `"device_memory_peak_mb": 4157.236992`

Now remove `safetensors` (which was installed with v4.30):

```
pip uninstall safetensors
python benchmark_model.py --benchmark_id="${BENCHMARK_ID}" --device=gpu --output_path="${OUTPUT_PATH}" --iterations=5
```

Benchmark output will show `"device_memory_peak_mb": 1591.090432`

### Expected behavior

Device peak memory usage should not have increased by 2.5x.

### PR DESCRIPTION
**Pull request by Rocketknight1**
When we load safetensors files in TF, the entire safetensors state dict is materialized on GPU alongside the randomly initialized weights. This inflates our memory usage during loading a lot, up to about 2X - 2.5X the amount we actually need.

This PR grabs tensors iteratively from the underlying safetensors archives and assigns them. ~It's working for TF-formatted safetensors archives right now, and I'll add torch-format support next.~ Now supports PT and TF formatted archives! Load times still seem very quick for me locally, so I don't think this negatively impacts anything!

Fixes #24393

### CODE REVIEW

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (sgugger) [inline src/transformers/modeling_tf_pytorch_utils.py:266]: No we don't trust Python bool conversion magic and never will, so let's revert that change or add an or for the False test if it;s what you want to do here.

**Review by sgugger — APPROVED** [review]
It looks good to me, but I just got out of dental surgery and this might be the painkillers talking. I'd love for @amyeroberts to have a look before this gets merged.

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_pytorch_utils.py:266]: lol that is dedication to duty, go lie down and talk to the hallucinations for a bit, I promise not to break the codebase while you're gone!

**Comment by Rocketknight1** [comment]
This looks good in my testing! cc @Narsil for review and @amyeroberts for core maintainer review to save Sylvain from having to read code while he's distracted by the hallucinatory dental anaesthetic goblins 

To explain one detail: In general, all TF ops can take NumPy inputs. Often the "idiomatic" way to write TF code is to do preprocessing/once-off stuff in NumPy and just pass the result to TF. Since Safetensors have very efficient Numpy loading, I generally open them in `"np"` format and let TF handle any necessary conversions.

The one exception is when loading a PyTorch state dict archive. The reason here is that PyTorch weights often need to be transposed to load them into TF, and NumPy transposes on CPU are much slower than TF transposes on GPU. Model loading was several seconds slower when I passed `"np"` format archives to the PyTorch state dict crossloading function, but with GPU transposes this PR has almost no impact on performance, while hugely reducing peak memory usage.

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for digging into and fixing this :)

File: src/transformers/modeling_tf_utils.py
Comment (amyeroberts) [reply src/transformers/modeling_tf_utils.py:2856]: ^ I see you answered above before I finalised my review :)

Comment (Narsil) [inline src/transformers/modeling_tf_utils.py:2856]: Does this have any impact? 

keeping everything `tf` in a `tf` files seems easier imho.

File: src/transformers/modeling_tf_utils.py
Comment (Narsil) [inline src/transformers/modeling_tf_utils.py:2898]: I'm slightly horrified to see a handle be called a dict, but it's a clever hack to not rewrite too many things.

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (Narsil) [inline src/transformers/modeling_tf_pytorch_utils.py:335]: Is is ok to change type here ? Shouldn't it be `array.shape`?

File: src/transformers/modeling_tf_utils.py
Comment (Narsil) [inline src/transformers/modeling_tf_utils.py:1028]: Does this automatically send the weight to GPU in `tf`?

File: src/transformers/modeling_tf_utils.py
Comment (Narsil) [inline src/transformers/modeling_tf_utils.py:1002]: Why is it `"np"` here ?

You are already using 1 tensor at a time, so having  `"np"` or tf shouldn't chnage anything, no ?

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_pytorch_utils.py:335]: Oh woah, this is a bug, you're right. Sorry!

Comment (sgugger) [reply src/transformers/modeling_tf_pytorch_utils.py:335]: Good catch, there is `shape` missing here.

File: src/transformers/modeling_tf_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:2898]: I'll rename it here, and hopefully the docstring on the actual state dict loading function is clear enough!

File: src/transformers/modeling_tf_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:1028]: Yep - `weight` is already a GPU tensor at this point, and `weight_value` can be anything - NumPy, CPU tensor or GPU tensor. It will always correctly update `weight` in-place.

File: src/transformers/modeling_tf_pytorch_utils.py
Comment (Rocketknight1) [reply src/transformers/modeling_tf_pytorch_utils.py]: Fixed!

File: src/transformers/modeling_tf_utils.py
Comment (Narsil) [reply src/transformers/modeling_tf_utils.py:1002]: Fair enough, I just thought seeing `"tf"` within a `tf_modeling` made more sense.  (and since it's required for the GPU transpose ops..)

**Review by Narsil — COMMENTED** [review]
LGTM ! 

I made some comments, mostly to try and keep consistent across the file (if possible).
But nothing blocking on my side.

**Comment by Rocketknight1** [comment]
Yeah, actually, I thought about it a little more - using `np` is idiomatic for TF, but it's a bit messy if the PyTorch crossloading uses `tf` for fast GPU transposes and the others use `np`, especially when it doesn't really matter. I'll use `tf` globally for consistency!

**Comment by Rocketknight1** [comment]
Just ran the slow tests for a few models that load safetensor checkpoints and this still seems good

**Comment by Rocketknight1** [comment]
Finished slow testing - performance and memory usage look good!

**Comment by Rocketknight1** [comment]
Everything looks good at this point - any objections to merging?

**Pull request merged**
